web/src/unread.ts

```diff
@@ -26,6 +26,7 @@ import * as util from "./util";
 // for more details on how this system is designed.
 
 export let old_unreads_missing = false;
+export let first_unread_unmuted_message_id = Number.POSITIVE_INFINITY;
```

Comment:
This global logic seems confusing and potentially wrong -- what updates it to be a higher number when messages are marked as read?

In any case, we should be able to just walk the data structures to get the answer with a function. But also ... I think ` narrow_state.get_first_unread_info` likely does exactly what we need here?

Reply:
> This global logic seems confusing and potentially wrong -- what updates it to be a higher number when messages are marked as read?

Oops! I added a change to reset it at the start so that it always capture the minimum of current unread counts.

```diff
diff --git a/web/src/unread.ts b/web/src/unread.ts
index 0f6fc29999..dfb0cd479c 100644
--- a/web/src/unread.ts
+++ b/web/src/unread.ts
@@ -903,6 +903,9 @@ export type FullUnreadCountsData = {
 // should strive to keep it free of side effects on globals or DOM.
 export function get_counts(): FullUnreadCountsData {
     const update_first_unmuted_message_id = true;
+    // Reset the first_unread_unmuted_message_id, to ensure it is always capture the
+    // minimum of current unread messages between topics and DMs.
+    first_unread_unmuted_message_id = Number.POSITIVE_INFINITY;
     const topic_res = unread_topic_counter.get_counts(update_first_unmuted_message_id);
     const pm_res = unread_direct_message_counter.get_counts(update_first_unmuted_message_id);
```


> In any case, we should be able to just walk the data structures to get the answer with a function.

Yeah, but since `unread_ui.update_unread_counts` is called quite frequently, we can always have `first_unread_unmuted_message_id` available which can be used for other purposes too.

> narrow_state.get_first_unread_info likely does exactly what we need here?

No, it needs message to be locally available in `message_store`.

Reply:
Note this value is basically what want for combined feed in https://github.com/zulip/zulip/pull/30008#discussion_r1597279862

Reply:
Is the concern with using `narrow_state.get_first_unread_info` avoiding import cycles?

Reply:
> Is the concern with using narrow_state.get_first_unread_info avoiding import cycles?

I replied with this above:
> No, it needs message to be locally available in message_store.

Since we don't have all the unreads in `message_store` we cannot get `first_unread_unmuted_message_id` using that method.